src/expr.jl

```diff
@@ -2,25 +2,12 @@
 # Builtin Expr type.
 
 istree(x::Expr) = true
-exprhead(e::Expr) = e.head
 
-operation(e::Expr) = expr_operation(e, Val{exprhead(e)}())
-arguments(e::Expr) = expr_arguments(e, Val{exprhead(e)}())
+operation(e::Expr) = e.head
+arguments(e::Expr) = e.args
 
-# See https://docs.julialang.org/en/v1/devdocs/ast/
-expr_operation(e::Expr, ::Union{Val{:call},Val{:macrocall}}) = e.args[1]
-expr_operation(e::Expr, ::Union{Val{:ref}}) = getindex
-expr_operation(e::Expr, ::Val{T}) where {T} = T
-
-expr_arguments(e::Expr, ::Union{Val{:call},Val{:macrocall}}) = e.args[2:end]
-expr_arguments(e::Expr, _) = e.args
-
-
-function similarterm(x::Expr, head, args, symtype = nothing; metadata = nothing, exprhead = exprhead(x))
-  expr_similarterm(head, args, Val{exprhead}())
-end
-
-
-expr_similarterm(head, args, ::Val{:call}) = Expr(:call, head, args...)
-expr_similarterm(head, args, ::Val{:macrocall}) = Expr(:macrocall, head, args...) # discard linenumbernodes?
-expr_similarterm(head, args, ::Val{eh}) where {eh} = Expr(eh, args...)
+function similarterm(x::Expr, head, args, symtype = nothing; metadata = nothing)
+  res = Expr(head)
+  res.args = args
```

Comment:
```
julia> c.args = [:f,2,3]
3-element Vector{Any}:
  :f
 2
 3

julia> c.args = (:f,2,3)
ERROR: MethodError: Cannot `convert` an object of type Tuple{Symbol, Int64, Int64} to an object of type Vector{Any}
```

Can you add a method dispatching on `args::Vector` and one that dispatches on `args::Any` and calls `collect`?

Reply:
oh, sure sure. I wasn't sure if Base had a better constructor for this, does it?

Reply:
`Expr(head, args...)` splatting is fast enough.

Reply:
Oh. also, really important to copy the vector!!
```
julia> using BenchmarkTools;

julia> function foo(v)
           for i in 1:100_000
               Expr(:call, v...)
           end
       end
foo (generic function with 1 method)

julia> function bar(v)
           for i in 1:100_000
               h = Expr(:call)
               h.args = v
           end
       end
bar (generic function with 1 method)

julia> v = [:f; collect(1:300)];

julia> foo(v);

julia> bar(v);

julia> @btime foo($v);
  493.611 ms (200000 allocations: 247.19 MiB)

julia> @btime bar($v);
  2.175 ms (200000 allocations: 7.63 MiB)
```

It's just setting args to be a reference, and it can be very dangerous as you're not actually constructing a new term if you do `ex = :(f(a,b,c)) ; similarterm(Expr, head(ex), arguments(ex))`.

Reply:
Okay! The args are splatted now.